statemachine/statemachine.go

```diff
@@ -0,0 +1,317 @@
+package statemachine
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/lytics/metafora"
+)
+
+type StateCode string
+
+const (
+	Runnable  StateCode = "runnable"  // Scheduled
+	Sleeping            = "sleeping"  // Scheduled, not running until time has elapsed
+	Completed           = "completed" // Terminal, not scheduled
+	Killed              = "killed"    // Terminal, not scheduled
+	Failed              = "failed"    // Terminal, not scheduled
+	Fault               = "fault"     // Scheduled, in error handling / retry logic
+	Paused              = "paused"    // Scheduled, not running
+)
+
+// Terminal states will never run and cannot transition to a non-terminal
+// state.
```

Comment:
probability a dumb idea...   But could we make these types of `classifications` configurable?